packages/@uppy/tus/src/index.ts

```diff
@@ -298,57 +357,83 @@ export default class Tus extends BasePlugin {
+        ) {
           // The navigator is offline, let's wait for it to come back online.
           if (!this.requests.isPaused) {
             this.requests.pause()
-            window.addEventListener('online', () => {
-              this.requests.resume()
-            }, { once: true })
+            window.addEventListener(
+              'online',
+              () => {
+                this.requests.resume()
+              },
+              { once: true },
+            )
           }
         }
         queuedRequest.abort()
         queuedRequest = {
           shouldBeRequeued: true,
-          abort () {
+          abort() {
             this.shouldBeRequeued = false
           },
-          done () {
-            throw new Error('Cannot mark a queued request as done: this indicates a bug')
+          done() {
+            throw new Error(
+              'Cannot mark a queued request as done: this indicates a bug',
+            )
           },
-          fn () {
+          fn() {
             throw new Error('Cannot run a queued request: this indicates a bug')
           },
         }
         return true
       }
 
       if (opts.onShouldRetry != null) {
-        uploadOptions.onShouldRetry = (...args) => opts.onShouldRetry(...args, defaultOnShouldRetry)
+        uploadOptions.onShouldRetry = (...args) =>
+          // @ts-expect-error TS does not pick up the != null check and also thinks ...args already
+          // contains the `next` argument we are about to add. It would require complex utility types
+          // to fix but it doesn't matter, the type is correct everywhere, it's just confused about this composition.
+          opts.onShouldRetry(...args, defaultOnShouldRetry)
       } else {
         uploadOptions.onShouldRetry = defaultOnShouldRetry
       }
 
-      const copyProp = (obj, srcProp, destProp) => {
+      const copyProp = (
+        obj: Record<string, unknown>,
+        srcProp: string,
+        destProp: string,
+      ) => {
         if (hasProperty(obj, srcProp) && !hasProperty(obj, destProp)) {
           // eslint-disable-next-line no-param-reassign
           obj[destProp] = obj[srcProp]
         }
       }
 
-      /** @type {Record<string, string>} */
-      const meta = {}
-      const allowedMetaFields = Array.isArray(opts.allowedMetaFields)
-        ? opts.allowedMetaFields
-        // Send along all fields by default.
+      // We can't use `allowedMetaFields` to index generic M
+      // and we also don't care about the type specifically here,
+      // we just want to pass the meta fields along.
+      const meta: Record<string, string> = {}
+      const allowedMetaFields =
+        Array.isArray(opts.allowedMetaFields) ?
+          opts.allowedMetaFields
+          // Send along all fields by default.
         : Object.keys(file.meta)
       allowedMetaFields.forEach((item) => {
-        meta[item] = file.meta[item]
+        meta[item] = file.meta[item] as string
```

Comment:
```suggestion
        meta[item] = file.meta[item] as string // todo this is a lie
```

Reply:
I agree that it may not be a `string`, but there is no todo here, as can be seen from the comment above.

Reply:
by asserting it `as string`, we are lying to typescript, while the truth is that meta is really a `Record<string, unknown>` so when we index it we don't know if there's really a string there. this could lead to bugs down the line. imo it's better to check whether it's actually a string and crash if not. OR change `metadata` in tus.UploadOptions to the type `Record<string,unknown>` (as well as tus's js implementation to allow that), however I'm not sure we want that.